config.yml

```diff
@@ -27,8 +27,11 @@ service:
   job_cleanup_interval: 300
   log_level: INFO
 
-connector_id: 'changeme'
-service_type: 'changeme'
+#connectors:
+#  changeme:
```

Comment:
+1 to this suggestion!

Reply:
>Is it an idea to make the connector id part of the object instead of the key/value pair key?

It will be easier to look up a connnector in code if we have `connector_id` as the key.

>Also could we lose the elasticsearch heading if we're only supporting api_key per connector?

The elasticsearch heading is for (possible) future support of per-connector configuration override. See @jedrazb 's suggestion [here](https://elastic.slack.com/archives/C01795T48LQ/p1688737399405279)

So we can easily take the `elasticsearch` config of a particular connector to override the global one.

Reply:
> It will be easier to look up a connnector in code if we have connector_id as the key.

Imo expressive YAML is better - we should value user experience over coding convenience when the inconvenience in code is low.

Another thing to consider - will it be easy to validate the config if user entered several connectors with same id in both approaches?

In the list-style approach proposed by Sander this validation is easier, in the hash-style approach one of the connector definitions can be eaten by YML parser OR probably generate a weird error message.

Reply:
> So we can easily take the elasticsearch config of a particular connector to override the global one.

Well, we also have a bit of a weird `elasticsearch` configuration TBH. It partially combines the default elasticsearch client configuration + our configuration that is not related to elasticsearch, but related to service config - it was just faster to implement it this way.

I think we need to detach settings like `max_concurrency` and such since they are not purely elasticsearch settings, these are framework settings.

Reply:
>Another thing to consider - will it be easy to validate the config if user entered several connectors with same id in both approaches?

We don't have any validation now. Connector service will raise an error and exit if the config is not a valid ymal. If we use hash-style, duplicate connector id will fail the ymal validation.

>In the list-style approach proposed by Sander this validation is easier, in the hash-style approach one of the connector definitions can be eaten by YML parser OR probably generate a weird error message.

From the validation perspective, I do think hash-style is easier, as we don't need to implement any validation, duplicate connector id will fail the yaml parsing.

>we should value user experience over coding convenience when the inconvenience in code is low.

I think this is a valid point. If the team think list is a better user experience I'm OK to change it.

>It partially combines the default elasticsearch client configuration + our configuration that is not related to elasticsearch, but related to service config

I agree that those configs under `elasticsearch.bulk` are not `elasticsearch` related, strictly speaking. But Iit will still help to override other `elasticsearch` settings in the future with `elasticsearch` heading, e.g. `request_timeout`.

Reply:
> I agree that those configs under elasticsearch.bulk are not elasticsearch related, strictly speaking. But Iit will still help to override other elasticsearch settings in the future with elasticsearch heading, e.g. request_timeout.

Yes, but it also would not make a lot of sense to allow overriding _all_ elasticsearch configs either. I feel like if we group these settings under `elasticsearch` key it's just gonna make it look like you can provide fully distinct Elasticsearch config per connector, but it honestly does not make a lot of sense. Or does it?

> We don't have any validation now. Connector service will raise an error and exit if the config is not a valid ymal. If we use hash-style, duplicate connector id will fail the ymal validation.

But how user-friendly is the error? Do you have an example message?